File: c/sorting/insertion_sort.c
#include <stdio.h>
#include "../benchmark.c"

void insertion_sort(int* array, int array_size) {
    // Start at the 2nd item (i=0)
    benchmark();
    for (size_t i = 1; i < array_size; i++) {
        int n = array[i];
        int j = i-1;

        // Continuously check previous until first.
        while (j >= 0 && array[j] > n) {
            array[j+1] = array[j];
            j--;
        }
        // Insert ahead:
        array[j+1] = n;
    }
    end_benchmark();
}

File: rust/src/sorting/insertion_sort.rs
use crate::benchmark::benchmark;

/*
*
* INSERTION SORT: SAFE
*
*/
fn insertion_sort(array: &mut Vec<u32>) {
    // Start with 2nd item:
    for i in 1..array.len() {
        let n = array[i];
        let mut j = i;

        // Check previous:
        while j > 0 && array[j-1] > n {
            array[j] = array[j-1];
            j-=1;
        }

        // Insert front:
        array[j] = n;
    }
}

/*
*
* INSERTION SORT: UNSAFE
*
*/
fn insertion_sort_unsafe(array: &mut Vec<u32>) {
    // Start with 2nd item:
    unsafe {
        for i in 1..array.len() {
            let n = *array.get_unchecked(i);
            let mut j = i;

            // Check previous:
            while j > 0 && *array.get_unchecked(j-1) > n {
                array[j] = *array.get_unchecked(j-1);
                j-=1;
            }

            // Insert front:
            array[j] = n;
        }
    }
}

/*
* start_sort: accepts 2D array and benchmarks per each iteration.
*/
pub fn do_benchmark(array: &mut Vec<Vec<u32>>, use_unsafe : bool) {
    for mut sub_array in array.iter_mut() {
        if use_unsafe {
            benchmark(1, || insertion_sort_unsafe(sub_array));
        } else {
            benchmark(1, || insertion_sort(sub_array));
        }
    }
}
